src/evo/deterministicmns.h

```diff
@@ -0,0 +1,606 @@
+    CDeterministicMN(const CDeterministicMN& mn, uint64_t _internalId) : CDeterministicMN(mn) {
+        // only non-initial values
+        assert(_internalId != std::numeric_limits<uint64_t>::max());
+        internalId = _internalId;
+    }
+
+    template <typename Stream>
+    CDeterministicMN(deserialize_type, Stream& s)
+    {
+        s >> *this;
+    }
+
+    uint256 proTxHash;
+    COutPoint collateralOutpoint;
+    uint16_t nOperatorReward;
+    CDeterministicMNStateCPtr pdmnState;
+
+public:
+    template <typename Stream, typename Operation>
+    inline void SerializationOp(Stream& s, Operation ser_action)
+    {
+        READWRITE(proTxHash);
+        READWRITE(VARINT(internalId));
+        READWRITE(collateralOutpoint);
+        READWRITE(nOperatorReward);
+        READWRITE(pdmnState);
+    }
+
+    template<typename Stream>
+    void Serialize(Stream& s) const { NCONST_PTR(this)->SerializationOp(s, CSerActionSerialize()); }
+
+    template<typename Stream>
+    void Unserialize(Stream& s) { SerializationOp(s, CSerActionUnserialize()); }
+
+    uint64_t GetInternalId() const;
+
+    std::string ToString() const;
+    void ToJson(UniValue& obj) const;
+};
+
+typedef std::shared_ptr<const CDeterministicMN> CDeterministicMNCPtr;
+
+class CDeterministicMNListDiff;
+
+template <typename Stream, typename K, typename T, typename Hash, typename Equal>
+void SerializeImmerMap(Stream& os, const immer::map<K, T, Hash, Equal>& m)
+{
+    WriteCompactSize(os, m.size());
+    for (typename immer::map<K, T, Hash, Equal>::const_iterator mi = m.begin(); mi != m.end(); ++mi)
+        Serialize(os, (*mi));
+}
+
+template <typename Stream, typename K, typename T, typename Hash, typename Equal>
+void UnserializeImmerMap(Stream& is, immer::map<K, T, Hash, Equal>& m)
+{
+    m = immer::map<K, T, Hash, Equal>();
+    unsigned int nSize = ReadCompactSize(is);
+    for (unsigned int i = 0; i < nSize; i++) {
+        std::pair<K, T> item;
+        Unserialize(is, item);
+        m = m.set(item.first, item.second);
+    }
+}
+
+// For some reason the compiler is not able to choose the correct Serialize/Deserialize methods without a specialized
+// version of SerReadWrite. It otherwise always chooses the version that calls a.Serialize()
+template<typename Stream, typename K, typename T, typename Hash, typename Equal>
+inline void SerReadWrite(Stream& s, const immer::map<K, T, Hash, Equal>& m, CSerActionSerialize ser_action)
+{
+    ::SerializeImmerMap(s, m);
+}
+
+template<typename Stream, typename K, typename T, typename Hash, typename Equal>
+inline void SerReadWrite(Stream& s, immer::map<K, T, Hash, Equal>& obj, CSerActionUnserialize ser_action)
+{
+    ::UnserializeImmerMap(s, obj);
+}
+
+class CDeterministicMNList
+{
+public:
+    typedef immer::map<uint256, CDeterministicMNCPtr> MnMap;
+    typedef immer::map<uint64_t, uint256> MnInternalIdMap;
+    typedef immer::map<uint256, std::pair<uint256, uint32_t> > MnUniquePropertyMap;
```

Comment:
What is the rationale behind the counter? I don't think that it's ever used.

If this map's purpose is to keep track of the properties to not accept duplicates and the keys inside a MN cannot be re-used between roles, then this counter should be always 1.

Reply:
Looks like it, yes.
But I'd rather leave this kind of refactoring for the end, when we'll have more tests covering unique properties.
It doesn't hurt to keep it for now.